Use cpp headers over system headers

sarahec opened the pull request:
This replaces system headers such as `<math.h>` with the corresponding C++ header (`<cmath>`) ni C++ code.

Each of the compiler suites includes its own c++ headers, and there's a risk of the system headers differing from these and causing subtle bugs.

sarahec commented:
The formatting script is sorting the headers and re-wrapped some lines in base.cc and double.cc. Please take a look.

pjaaskel commented on openasip/newlib-1.17.0/newlib/libc/sys/linux/linuxthreads/machine/i386/useldt.h:
```diff
@@ -19,9 +19,8 @@
-#include <stdlib.h>	/* For abort().  */
-
+#include <cstddef> /* For offsetof.  */
+#include <cstdlib> /* For abort().  */
```
Please avoid touching the newlib files (this is a copy from upstream which we might some day want to synchronize to a newer version so it's best to minimize the diff). Also, this is a C library.

sarahec replied:
I'll revert newlib. (Strange, my changes weren't supposed to touch the C files.)

pjaaskel commented on openasip/opset/base/base.cc:
```diff
@@ -1273,9 +1273,15 @@ OPERATION(EQF)
-        IO(3) = (!isnan(FLT(1)) && !isnan(FLT(2)) && FLT(1) == FLT(2)) ? 1 : 0;
+        IO(3) =
+            (!std::isnan(FLT(1)) && !std::isnan(FLT(2)) && FLT(1) == FLT(2))
+                ? 1
```
Yep, this doesn't look nice :) Can you describe the original style in clang-format options?

pjaaskel reviewed (changes requested):
The ? : style doesn't look like most of TCE style and newlib should be left with minimal modifications, otherwise LGTM.

pjaaskel commented on openasip/newlib-1.17.0/newlib/libc/sys/linux/linuxthreads/machine/i386/useldt.h:
```diff
@@ -19,9 +19,8 @@
    write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
    Boston, MA 02111-1307, USA.  */
 
-#include <stddef.h>	/* For offsetof.  */
```
Can you revert the newlib change?

sarahec replied:
Of course! Reverted.

pjaaskel commented on openasip/opset/base/base.cc:
```diff
@@ -1273,9 +1275,9 @@ OPERATION(EQF)
 TRIGGER
     if (BWIDTH(1) == FLT_WORD_SIZE && BWIDTH(2) == FLT_WORD_SIZE) {
-        IO(3) = (!isnan(FLT(1)) && !isnan(FLT(2)) && FLT(1) == FLT(2)) ? 1 : 0;
+      IO(3) = (!isnan(FLT(1)) && !isnan(FLT(2)) && FLT(1) == FLT(2)) ? 1 : 0;
```
This indent doesn't look right. We use 4 spaces, no tabs in OpenASIP.

pjaaskel commented on openasip/opset/base/double.cc:
```diff
@@ -134,7 +134,8 @@ END_OPERATION(DIVD)
 TRIGGER
-IO(3) = (!isnan(DBL(1)) && !isnan(DBL(2)) && DBL(1) == DBL(2)) ? 1 : 0;
+IO(3) = (!std::isnan(DBL(1)) && !std::isnan(DBL(2)) && DBL(1) == DBL(2)) ? 1
+                                                                         : 0;
```
This still has the weird newline before the ":".

sarahec replied:
Reverted from main, updated only the header

sarahec commented:
Changes in the force-pushed update:

1. Reverted all of newlib (copied over from main)
2. Reverted base.cc and double.cc (copied from main), then changed the header and added `using namespace std`. No other formatting changes applied.
3. Squashed and force-pushed the combined commits to make the merge cleaner. Here's the [diff link](https://github.com/cpc/openasip/pull/221/commits/a0b8a6fcaa7b46c131bbf4684974a38b07c2d3c1) for your review.

pjaaskel commented:
Did you run the tests? There seem to be a couple of cases which are compiled with the oacc's (incomplete C++ support), which doesn't yet have a stdlib.

```
<system_testing errors with Current shell environment. With revision a0b8a6fcaa7b46c131bbf4684974a38b07c2d3c1.>
[349637] ./procgen/hpu/test_hpu_embedded.testdesc: half-precision FPUs by generating and simulating a processor...FAIL (0m0.569s)
[349645] ./bintools/Compiler/RegressionTests/bool_return_value.testdesc: Test that boolean return values work...FAIL (0m0.436s)
Differences found against verification data are stored into difference.txt
</system_testing errors with Current shell environment. With revision a0b8a6fcaa7b46c131bbf4684974a38b07c2d3c1.>
```
Here's the info for how to run the tests if you haven't before: https://github.com/cpc/openasip/wiki/Contributor-Info

sarahec commented:
Thank you. I had been running tests from the test directory and missed that there's a test driver script. It's also flagging problems I hadn't seen earlier.

This may take some time, as compiletest isn't passing on a fresh checkout of the project.

sarahec commented:
Filed #223 which is blocking resolution of the above question.

sarahec commented:
I'm going to close this for now and come back when I have a reliable testing setup.
